### PR TITLE
[TwigBundle][FrameworkBundle] Fixed that Twig cache warmer is not being added as cache warmer

### PR DESCRIPTION
**Pull request by denniscoorn**
## Description

Since version 2.7 the Twig cache warmer is not being triggered during a `app/console cache:warmup`. The bug is introduced with commit https://github.com/symfony/symfony/commit/0d537c4e4d3da375f67b601f0ecc9f93c9ab4388 where the tag `kernel.cache_warmer` is removed from the service definition `twig.cache_warmer` and is added dynamically by the `ExtensionPass`. 

The problem lies in the fact that the `AddCacheWarmerPass` of the `FrameworkBundle` has already been processed on the moment the `ExtensionPass` adds the tag to the `twig.cache_warmer` service. Therefore the `twig.cache_warmer` is not being picked up by the `AddCacheWarmerPass` and will not be used during an `app/console cache:warmup`.

To verify my hypothesis I made a small change in `AppKernel` and placed `TwigBundle`above `FrameworkBundle`, to force a different processing order of compiler passes and that fixed the problem.

**Steps to reproduce**
- Install a fresh copy of symfony `symfony new test`
- Run `app/console cache:warmup`

_Notice that the directory `app/cache/dev/twig` doesn't exists_.
## Solution

My suggested fix is small but effective and consists of adding the argument `PassConfig::TYPE_OPTIMIZE` to the `addCompilerPass` method call when adding the `AddCacheWarmerPass` to the container. This ensures that other bundles were able to add tags to there service definitions, regarding cache warmers, before the `AddCacheWarmerPass` is being called.

| Q | A |
| --- | --- |
| Bug fix? | yes |
| New feature? | no |
| BC breaks? | no |
| Deprecations? | no |
| Tests pass? | yes |
| Fixed tickets | #15982 |
| License | MIT |
| Doc PR | - |

### CODE REVIEW

**Comment by fabpot** [comment]
Can you add a test to avoid any future regression?

**Comment by denniscoorn** [comment]
@fabpot Just added a test to avoid any future regression. I doubted a little bit about where I should add this test, because the fix is applied on the `FrameworkBundle` but the issue is closely related to the `TwigBundle`. Eventually I added a test to the `TwigBundle` because the main goal of the test is to ensure that the `TemplateCacheCacheWarmer` is being added to the cache warmer.

**Comment by denniscoorn** [comment]
After investigating what causes the Travis CI build to fail I realized that the test I added is never going to pass when the `TwigBundle` tests are executed independently. Therefore I removed the previous committed `TemplateCacheCacheWarmerTest` from the `TwigBundle` and added a more generic `CacheWarmerTest` to the `FrameworkBundle`.

I hope I did everything as it should be. If not, please comment! :-)

**Comment by stof** [comment]
there is an issue. This compiler pass should still run **before** the optimization passes, otherwise it could break optimizations (if you have circular references between your cache warmer and the CacheWarmerAggregate, they won't be detected anymore for instance because the circular reference will be created only after the validation. and there might be other broken cases). So -1 for the current patch.

IMO, what is needed is a way to set a priority between all compiler passes running before optimizations (because almost every compiler pass provided by bundles must actually run before optimizing the container)

**Comment by xabbuh** [comment]
What would actually happen when added the `kernel.cache_warmer` tag even if the `templating` service is not registered? Would that be an issue at all?

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/CacheWarmer/CacheWarmer.php [hunk 1]
```diff
@@ -0,0 +1,20 @@
...
+    {
+        $filesystem = new Filesystem();
+        $filesystem->mkdir($cacheDir.'/cache_warmer');
+    }
+
+    public function isOptional()
+    {
+        return true;
+    }
+}
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/CacheWarmer/CacheWarmer.php]: missing newline

File: src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/DependencyInjection/Compiler/CacheWarmerPass.php [hunk 2]
```diff
@@ -0,0 +1,16 @@
...
+
+class CacheWarmerPass implements CompilerPassInterface
+{
+    public function process(ContainerBuilder $container)
+    {
+        if ($container->hasDefinition('cache_warmer_test.cache_warmer')) {
+            $container->getDefinition('cache_warmer_test.cache_warmer')->addTag('kernel.cache_warmer');
+        }
+    }
+}
```
Comment (Tobion) [inline src/Symfony/Bundle/FrameworkBundle/Tests/Functional/Bundle/TestBundle/DependencyInjection/Compiler/CacheWarmerPass.php]: newline

**Comment by fabpot** [comment]
Closing for the reasons explained by @stof

**Pull request closed without merging**